third_party/nvfuser/csrc/multidevice/aggregate_dag.cpp

```diff
@@ -0,0 +1,326 @@
+            addInput(av);
+        }
+      }
+    }
+  }
+
+  // Build the AggregateExpr IR of the AggregateDag
+  // An AggregateExpr is created for each Cluster of the MultiClusterFusion
+  void AggregateDag::buildAExpr() {
+    for (auto cluster : MGFusion_->clusters()) {
+
+    // the inputs (resp. outputs) of AggregateExpr are the 
+    // cluster's consumed (resp. produced) vals
+      auto output_container = FilterAggregateValContainer(vals_, 
+                                      [](auto val){return val->isValOutputOfCluster();},
+                                      nullptr, cluster);
+      auto input_container = FilterAggregateValContainer(vals_, 
+                                      [](auto val){return val->isValInputOfCluster();},
+                                      nullptr, cluster);
+
+      auto aExpr = IrBuilder::create<AggregateExpr>(this->as<IrContainer>(), cluster);
+
+      for (auto v: output_container){
+        v->setDefinition(aExpr);
+      }
+    }
+  }
+
+  // Build the SendRecv IR of the AggregateDag
+  // A SendRecv is created for each pair of AggregateVals which Vals match
+  // and which are not inputs/outputs of the AggregateDag
+  void AggregateDag::buildSendRecv() {
+    //select all AggregateVal that are not global I/O of the AggregateDag
+    decltype(vals_) internal_aVals;
+    std::copy_if(vals_.begin(),
+                vals_.end(),
+                std::inserter(internal_aVals, internal_aVals.end()),
+                [&](auto val){
+                  return std::count(inputs().begin(), inputs().end(), val)
+                      + std::count(outputs().begin(), outputs().end(), val) == 0;}
+                );
+
+    // Builds the set of original Vals represented in internal_aVals
+    std::unordered_set<Val*> internal_Vals;
+    std::transform(
+        internal_aVals.cbegin(),
+        internal_aVals.cend(),
+        std::inserter(internal_Vals, internal_Vals.end()),
+        [](Val* val){return val->as<AggregateVal>()->getOriginalVal();});
+
+    for (auto val: internal_Vals) {
+      auto srcs = FilterAggregateValContainer(internal_aVals, 
+                        [](auto val){return val->isValOutputOfCluster();}, val);
+      auto dsts = FilterAggregateValContainer(internal_aVals, 
+                        [](auto val){return val->isValInputOfCluster();}, val);
+      TORCH_INTERNAL_ASSERT(
+            std::size(dsts) > 0,
+            "WARN: Try to define SendRecv IR with no destinations");
+      TORCH_INTERNAL_ASSERT(
+            std::size(srcs) == 1,
+            "WARN: Try to define SendRecv IR with a number of sources not equal to 1");
+      auto src = (*srcs.begin())->as<AggregateVal>();
+      for (auto dst: dsts){
+        auto sr = IrBuilder::create<SendRecv>(this->as<IrContainer>(),
+                                              dst->as<AggregateVal>(),
+                                              src);
+        dst->setDefinition(sr);
```

Comment:
Can this be changed as well so that creating `SendRecv` automatically "defines" `dst`?

Reply:
I see no way to infer what dst is because there could be several dst (here we iterate over dsts which can contains several elements).
I'd rather leave it like this for now, if ok with you. Also, this will change in the next PR

Reply:
I'm not talking about automatically inferring the dependencies, but it's just a mechanical matter. This is effectively the same comment about `buildAExpr()`, which also explicitly calls `setDefinition()`.

Reply:
Got it, I remove `setDefinition` here also